src/webR/utils-r.ts

```
@@ -61,3 +61,42 @@ export function parseEvalBare(code: string, env: WebRData): RObject {
     unprotect(prot.n);
   }
 }
+
+let evalBodyPtr: EmPtr | undefined;
+let evalHandlerPtr: EmPtr | undefined;
+let evalErrorMessage: string | undefined;
+
+function evalCallBody(data: EmPtr): RPtr {
+  const call = Module.getValue(data, 'i32');
+  const env = Module.getValue(data + 4, 'i32');
+  return Module._Rf_eval(call, env);
+}
+
+function evalCallHandler(cond: RPtr, _: EmPtr) {
+  evalErrorMessage = RObject.wrap(cond).get(1).toString();
+  return 0;
+}
+
+export function evalCall(call: RHandle, env: RHandle): RPtr {
+  if (!evalBodyPtr) {
+    evalBodyPtr = Module.addFunction(evalCallBody, 'ii');
+  }
+  if (!evalHandlerPtr) {
+    evalHandlerPtr = Module.addFunction(evalCallHandler, 'iii');
+  }
+
+  const dataPtr = Module._malloc(8);
+
+  Module.setValue(dataPtr, handlePtr(call), '*');
+  Module.setValue(dataPtr + 4, handlePtr(env), '*');
+
+  // In case of an error ensure that the stack is unwound only as far as this
+  // function, rather than up to top level, before throwing a JS error.
+  const res = Module._R_tryCatchError(evalBodyPtr, dataPtr, evalHandlerPtr, 0);
```

The stack might be unwound with other events than errors, for instance a restart invokation or a `Q` event in `browser()`.

It would be safer to use a general purpose longjump catcher like `R_UnwindProtect()`. Unfortunately there's no way to extract errors from longjump tokens, so we'd still need the trycatch to convert R errors to JS errors.

I guess the simplest way is to call `R_tryCatchError()` from an `R_ExecWithCleanup()` context to detect other kinds of longjumps and throw a generic error in that case.

The more accurate way would be to use `R_UnwindProtect()`, throw a special javascript exception that wraps the longjump token in case of early exit, and then resumes the jump from `readConsole()`. Then we don't need to capture R errors I think. This would be the job of `captureR()` as before.

The problem with this approach is that it's incompatible with catch-all blocks though. The longjump exception should not be caught on the way to `readConsole()`.

---

Thanks for this great suggestion, I had not properly understood `R_UnwindProtect()` until now but it really looks like the right tool for the job. I've reimplmented things on this branch to try and match your third suggestion - using `R_UnwindProtect()` with a custom javascript exception.

As a quick confirmation of how I understand what should be happening:
When an error (or other stack unwinding event) occurs during `safeEval()` a JS exception of type  `UnwindProtectException` is created. The custom exception takes a copy of the unwind protect token and is then thrown. The exception is passed up the call stack as it unwinds up to `readConsole()`, somtimes the exception is caught and re-thrown, where necessary. In `readConsole()` the exception is caught and `R_ContinueUnwind()` is invoked using the associated token, which then finally proceeds with the `NORET` `longjmp`.

With this change, R error conditions are no longer caught and converted into JS exceptions as part of `saveEval()`. Instead, they are written to `stderr` as normal. When evaluating an R call during a channel request, but not using `captureR` machinery, the associated promise on the main thread is rejected with a generic message[1] and any errors can be observed by consuming `stderr` messages from the output queue.

This seems to all work, but I appriciate another pair of eyes taking a look to make sure I have understood it all correctly and that the implemenation looks correct.

[1] For the moment the generic message has been set to "A non-local transfer of control occured during evaluation".

---

Your description matches my expectations about how this works.

I'll take another look in the morning with a fresher set of eyes :)